src/embedding/kd.ts

```diff
@@ -0,0 +1,752 @@
+
+    /**
+     * Type-guard checking whether object is an instance of {@link SetSizeCommandData}.
+     *
+     * @param obj - object to test
+     */
+    export function isSetSizeCommandData(obj: any): obj is SetSizeCommandData {
+        return getEventType(obj) === GdcKdCommandType.SetSize;
+    }
+
+    export interface INoPermissionsBody {
+        /**
+         * the 'data' section contains information about whether view or edit permissions are missing
+         */
+        reason: string;
+    }
+
+    /**
+     * This event is emitted When User does not have permissions to view or edit the content
+     */
+    export type NoPermissionsEventData = IKdGdcMessageEnvelope<
+        GdcKdEventType.NoPermissions,
+        INoPermissionsBody
+    >;
+
+    /**
+     * Type-guard checking whether object is an instance of {@link NoPermissionsEventData}.
+     *
+     * @param obj - object to test
+     */
+    export function isNoPermissionsEventData(obj: any): obj is NoPermissionsEventData {
+        return getEventType(obj) === GdcKdEventType.NoPermissions;
+    }
+
+    export interface IDashboardObjectMeta {
+        /**
+         * Client id - Each client has an identifier unique within the domain
+         *
+         * Note: use the combination of the data product ID and client ID instead of the project ID
+         */
+        client?: string;
+
+        /**
+         * object id
+         */
+        dashboardId: string;
+        /**
+         * Project id
+         */
+        project: string;
+        /**
+         * dashboard identifier
+         */
+        dashboard: string;
+
+        /**
+         * dashboard title - this is what users see in KD top bar (if visible)
+         */
+        title: string;
+    }
+
+    export type IDashboardBody = IKdAvailableCommands & IDashboardObjectMeta;
+
+    /**
+     * Data type of event that was emit when a dashboard has been created and saved.
+     */
+    export type IDashboardCreatedData = IKdGdcMessageEnvelope<
+        GdcKdEventType.DashboardCreated,
+        IDashboardBody
+    >;
+
+    /**
+     * Type-guard checking whether an object is an instance of {@link isDashboardCreatedData}
+     *
+     * @param obj - object to test
+     */
+    export function isDashboardCreatedData(obj: any): obj is IDashboardCreatedData {
+        return getEventType(obj) === GdcKdEventType.DashboardCreated;
+    }
+
+    /**
+     * Data type of event that was emit when the content is fully loaded,
+     * and the user has permissions to access the dashboard.
+     */
+    export type IDashboardLoadedData = IKdGdcMessageEnvelope<
+        GdcKdEventType.DashboardLoaded,
+        IDashboardBody
+    >;
+
+    /**
+     * Type-guard checking whether an object is an instance of {@link isDashboardLoadedData}
+     *
+     * @param obj - object to test
+     */
+    export function isDashboardLoadedData(obj: any): obj is IDashboardLoadedData {
```

Comment:
I might be missing something, but I see we have this type-guard alot for event from GDC to outer app.
I mean this is sent by GDC itself, we're already aware of what type the event is so why do we need this check?

Cc @lanhq

Reply:
Currently, this type-guard is unused. But it depend on the situation, it might be needed in future. The typing here is try to define it detail as much as possible.

Reply:
@lanhq, does this unuse happen to both AD and KD?
I think we should not add any unuse for now.

Reply:
yes. it happen on AD too. If so, i will remove these unuse on AD.

Reply:
I removed, only keep type-guard is using on KD